File: generate_sitemap.py
#!/usr/bin/env python3
"""
Script para gerar sitemap.xml automaticamente
para o projeto 365 de Graça & Adoração
"""

import os
from datetime import datetime
from pathlib import Path
import html

# Configurações
BASE_URL = "https://365-de-graca-e-adoracao.vercel.app"
OUTPUT_FILE = "sitemap.xml"
ROOT_DIR = Path(__file__).parent

# Prioridades por tipo de página
PRIORITIES = {
    'index': '1.0',
    'sobre': '0.9',
    'contexto': '0.9',
    'genesis_index': '0.8',
    'estudos': '0.7',
    'templates': '0.5',
}

# Frequência de atualização
CHANGEFREQ = {
    'index': 'weekly',
    'sobre': 'monthly',
    'contexto': 'monthly',
    'genesis_index': 'weekly',
    'estudos': 'monthly',
    'templates': 'yearly',
}

def get_page_info(filepath):
    """Determina prioridade e changefreq baseado no caminho"""
    path_str = str(filepath)
    
    if path_str == 'index.html':
        return PRIORITIES['index'], CHANGEFREQ['index']
    elif '/sobre/' in path_str:
        return PRIORITIES['sobre'], CHANGEFREQ['sobre']
    elif '/contexto/' in path_str:
        return PRIORITIES['contexto'], CHANGEFREQ['contexto']
    elif 'genesis/index.html' in path_str:
        return PRIORITIES['genesis_index'], CHANGEFREQ['genesis_index']
    elif '/estudos/' in path_str and not path_str.endswith('genesis-'):
        return PRIORITIES['estudos'], CHANGEFREQ['estudos']
    else:
        return PRIORITIES['templates'], CHANGEFREQ['templates']

def find_html_files():
    """Encontra todos os arquivos HTML relevantes"""
    html_files = []
    
    # Diretórios para incluir
    include_dirs = [
        '.',
        'sobre',
        'contexto',
        '01-principio',
        '02-pentateuco/genesis',
        '02-pentateuco/genesis/estudos',
    ]
    
    for dir_path in include_dirs:
        full_path = ROOT_DIR / dir_path
        if full_path.exists():
            for html_file in full_path.glob('*.html'):
                # Ignorar arquivos temporários e templates
                if not html_file.name.startswith('_') and html_file.name != 'template.html':
                    rel_path = html_file.relative_to(ROOT_DIR)
                    html_files.append(rel_path)
    
    return sorted(html_files)

def generate_sitemap():
    """Gera o arquivo sitemap.xml"""
    html_files = find_html_files()
    lastmod = datetime.now().strftime('%Y-%m-%d')
    
    # Início do XML
    xml_content = ['<?xml version="1.0" encoding="UTF-8"?>']
    xml_content.append('<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">')
    
    # Adicionar cada URL
    for filepath in html_files:
        # Converter caminho do arquivo para URL
        url_path = str(filepath).replace('\\', '/')
        if url_path == 'index.html':
            url = BASE_URL + '/'
        elif url_path.endswith('/index.html'):
            url = BASE_URL + '/' + url_path.replace('/index.html', '/')
        else:
            url = BASE_URL + '/' + url_path
        
        # Escapar caracteres especiais XML
        url = html.escape(url, quote=True)
        
        priority, changefreq = get_page_info(filepath)
        
        xml_content.append('  <url>')
        xml_content.append(f'    <loc>{url}</loc>')
        xml_content.append(f'    <lastmod>{lastmod}</lastmod>')
        xml_content.append(f'    <changefreq>{changefreq}</changefreq>')
        xml_content.append(f'    <priority>{priority}</priority>')
        xml_content.append('  </url>')
    
    # Fim do XML
    xml_content.append('</urlset>')
    
    # Escrever arquivo
    output_path = ROOT_DIR / OUTPUT_FILE
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write('\n'.join(xml_content))
    
    print(f"✅ Sitemap gerado com sucesso!")
    print(f"📄 Arquivo: {output_path}")
    print(f"🔗 {len(html_files)} páginas incluídas")
    
    return len(html_files)

if __name__ == '__main__':
    generate_sitemap()

File: generate-sitemap.js
/**
 * generate-sitemap.js
 * Gera sitemap.xml automaticamente para 365gracaeadoracao.com
 * Execute: node generate-sitemap.js
 * Roda automaticamente no build da Vercel via package.json "build" script
 */

const fs   = require('fs');
const path = require('path');

const BASE_URL  = 'https://365gracaeadoracao.com';
const TODAY     = new Date().toISOString().split('T')[0]; // YYYY-MM-DD
const OUT_FILE  = path.join(__dirname, 'sitemap.xml');

// Diretórios e arquivos a ignorar
const IGNORE_DIRS  = new Set(['.git', 'node_modules', 'assets', '__pycache__']);
const IGNORE_FILES = new Set(['sitemap.xml', 'sitemap-visual.html', '404.html']);

// ── Regras de prioridade e frequência ────────────────────────────────────────
function getUrlMeta(relPath) {
  const p = relPath.replace(/\\/g, '/');

  // Home
  if (p === 'index.html' || p === '') {
    return { priority: '1.0', changefreq: 'weekly' };
  }

  // Índices de seção (index.html dentro de pastas)
  if (p.endsWith('/index.html') || p.endsWith('\\index.html')) {
    return { priority: '0.9', changefreq: 'weekly' };
  }

  // Capítulos do NT aprofundados (08-novo-testamento/.../capitulos/capitulo-XX.html)
  if (p.includes('08-novo-testamento') && p.includes('/capitulos/capitulo-')) {
    return { priority: '0.9', changefreq: 'monthly' };
  }

  // Salmos
  if (p.includes('/salmos/capitulos/salmo-')) {
    return { priority: '0.9', changefreq: 'monthly' };
  }

  // Capítulos do AT
  if (p.includes('/capitulos/capitulo-')) {
    return { priority: '0.8', changefreq: 'monthly' };
  }

  // Personagens, estudos, artigos estratégicos
  if (p.includes('/personagens/') || p.includes('/estudos/') || p.includes('/como-estudar/')) {
    return { priority: '0.8', changefreq: 'monthly' };
  }

  // Mapas, timeline, linha do tempo
  if (p.includes('/mapas/') || p.includes('/timeline/') || p.includes('/linha-do-tempo/')) {
    return { priority: '0.7', changefreq: 'monthly' };
  }

  // Demais páginas
  return { priority: '0.6', changefreq: 'monthly' };
}

// ── Coletar todos os arquivos HTML ────────────────────────────────────────────
function getAllHtmlFiles(dir, base, results = []) {
  let entries;
  try { entries = fs.readdirSync(dir); } catch { return results; }

  for (const entry of entries) {
    if (IGNORE_DIRS.has(entry)) continue;

    const fullPath = path.join(dir, entry);
    const stat     = fs.statSync(fullPath);

    if (stat.isDirectory()) {
      getAllHtmlFiles(fullPath, base, results);
    } else if (entry.endsWith('.html') && !IGNORE_FILES.has(entry)) {
      results.push(path.relative(base, fullPath));
    }
  }
  return results;
}

// ── Converter caminho em URL ──────────────────────────────────────────────────
function toUrl(relPath) {
  let url = relPath.replace(/\\/g, '/');

  // Remover index.html final (URL limpa)
  url = url.replace(/\/index\.html$/, '/').replace(/^index\.html$/, '');

  // Remover extensão .html
  url = url.replace(/\.html$/, '');

  // Codificar caracteres especiais (acentos, espaços)
  url = url.split('/').map(segment => encodeURIComponent(decodeURIComponent(segment))).join('/');

  return BASE_URL + '/' + url;
}

// ── Gerar XML ─────────────────────────────────────────────────────────────────
function buildXml(files) {
  const urlEntries = files.map(relPath => {
    const url  = toUrl(relPath);
    const meta = getUrlMeta(relPath);
    return `  <url>
    <loc>${url}</loc>
    <lastmod>${TODAY}</lastmod>
    <changefreq>${meta.changefreq}</changefreq>
    <priority>${meta.priority}</priority>
  </url>`;
  });

  return `<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9"
        xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
        xsi:schemaLocation="http://www.sitemaps.org/schemas/sitemap/0.9
        http://www.sitemaps.org/schemas/sitemap/0.9/sitemap.xsd">
${urlEntries.join('\n')}
</urlset>`;
}

// ── Validação básica de XML ───────────────────────────────────────────────────
function validateXml(xml) {
  // Checar & não escapado
  const badAmpersand = /&(?!amp;|lt;|gt;|quot;|apos;|#\d+;|#x[\da-fA-F]+;)/g;
  const matches = xml.match(badAmpersand);
  if (matches) {
    console.warn(`⚠️  Encontrados ${matches.length} & não escapados — corrija os nomes de arquivo.`);
  }
  return !matches;
}

// ── Main ──────────────────────────────────────────────────────────────────────
(function main() {
  const base  = __dirname;
  const files = getAllHtmlFiles(base, base);

  // Ordenar: home primeiro, depois índices, depois capítulos
  files.sort((a, b) => {
    const pa = getUrlMeta(a).priority;
    const pb = getUrlMeta(b).priority;
    if (pa !== pb) return parseFloat(pb) - parseFloat(pa); // maior prioridade primeiro
    return a.localeCompare(b);
  });

  const xml = buildXml(files);

  if (!validateXml(xml)) {
    console.error('❌ sitemap.xml contém erros de XML. Verifique os nomes de arquivo.');
    process.exit(1);
  }

  fs.writeFileSync(OUT_FILE, xml, 'utf8');

  // Estatísticas
  const byPriority = {};
  files.forEach(f => {
    const p = getUrlMeta(f).priority;
    byPriority[p] = (byPriority[p] || 0) + 1;
  });

  console.log('✅ sitemap.xml gerado com sucesso!');
  console.log(`   Total de URLs: ${files.length}`);
  console.log(`   Data: ${TODAY}`);
  console.log(`   Arquivo: ${OUT_FILE}`);
  Object.entries(byPriority).sort((a,b) => b[0]-a[0]).forEach(([p, n]) => {
    console.log(`   Prioridade ${p}: ${n} URLs`);
  });
})();
